### PR TITLE
[Feature] Support Cluster Snapshot Backup: SQL Interface and meta data (part 1)

### PR DESCRIPTION
**Pull request by srlch**
## What I'm doing:
This is part 1 for support backup cluster snapshot:

Introduce SQL Interface for TURN ON and TRUN OFF the automated cluster snapshot task:
```
ADMIN SET AUTOMATED CLUSTER SNAPSHOT ON
[STORAGE VOLUME <storage_volume_name>]

ADMIN SET AUTOMATED CLUSTER SNAPSHOT OFF
```

Fixes #53867
https://github.com/StarRocks/starrocks/issues/53867

## What type of PR is this:

- [ ] BugFix
- [x] Feature
- [ ] Enhancement
- [ ] Refactor
- [ ] UT
- [ ] Doc
- [ ] Tool

Does this PR entail a change in behavior?

- [ ] Yes, this PR will result in a change in behavior.
- [x] No, this PR will not result in a change in behavior.

If yes, please specify the type of change:

- [ ] Interface/UI changes: syntax, type conversion, expression evaluation, display information
- [ ] Parameter changes: default values, similar parameters but with different default values
- [ ] Policy changes: use new policy to replace old one, functionality automatically enabled
- [ ] Feature removed
- [ ] Miscellaneous: upgrade & downgrade compatibility, etc.

## Checklist:

- [x] I have added test cases for my bug fix or my new feature
- [ ] This pr needs user documentation (for new or modified features or behaviors)
  - [ ] I have added documentation for my new feature or new function
- [ ] This is a backport pr

## Bugfix cherry-pick branch check:
- [x] I have checked the version labels which the pr will be auto-backported to the target branch
  - [x] 3.4
  - [ ] 3.3
  - [ ] 3.2
  - [ ] 3.1
  - [ ] 3.0

### CODE REVIEW

File: fe/fe-core/src/main/java/com/starrocks/sql/parser/StarRocks.g4 [hunk 1]
```diff
@@ -734,6 +735,10 @@ syncStatement
     : SYNC
     ;
 
+adminSetAutomatedSnapshotStatement
+    : ADMIN SET AUTOMATED CLUSTER SNAPSHOT (ON (STORAGE VOLUME svName=identifier)? | OFF)
+    ;
```
Comment (xiangguangyxg) [inline fe/fe-core/src/main/java/com/starrocks/sql/parser/StarRocks.g4:740]:
```suggestion
adminSetAutomatedSnapshotOnStatement
    : ADMIN SET AUTOMATED CLUSTER SNAPSHOT ON (STORAGE VOLUME svName=identifier)?
    ;

adminSetAutomatedSnapshotOffStatement
    : ADMIN SET AUTOMATED CLUSTER SNAPSHOT OFF
    ;
```

File: fe/fe-core/src/main/java/com/starrocks/lake/snapshot/ClusterSnapshotMgr.java [hunk 2]
```diff
@@ -0,0 +1,198 @@
...
+    private ClusterSnapshot automatedSnapshot = null;
+    @SerializedName(value = "historyAutomatedSnapshotJobs")
+    private List<ClusterSnapshotJob> historyAutomatedSnapshotJobs = Lists.newArrayList();
+
+    private BlobStorage remoteStorage;
+    private String locationWithServiceId;
+
+    public ClusterSnapshotMgr() {}
+
+    public void addAutomatedSnapshotRequest(String storageVolumeName, boolean isReplay) {
```
Comment (xiangguangyxg) [inline fe/fe-core/src/main/java/com/starrocks/lake/snapshot/ClusterSnapshotMgr.java:64]:
```suggestion
    // Turn on automated snapshot, use stmt for extension in future
    public void setAutomatedSnapshotOn(AdminSetAutomatedSnapshotOnStmt stmt) {
```

File: fe/fe-core/src/main/java/com/starrocks/lake/snapshot/ClusterSnapshotMgr.java [hunk 3]
```diff
@@ -0,0 +1,198 @@
...
+
+    public ClusterSnapshot getAutomatedSnapshot() {
+        return automatedSnapshot;
+    }
+
+    public boolean containsAutomatedSnapshot() {
+        return getAutomatedSnapshot() != null;
+    }
+
+    public boolean containAutomatedSnapshotRequest() {
```
Comment (xiangguangyxg) [inline fe/fe-core/src/main/java/com/starrocks/lake/snapshot/ClusterSnapshotMgr.java:121]:
```suggestion
    public boolean isAutomatedSnapshotOn() {
```

File: fe/fe-core/src/main/java/com/starrocks/sql/parser/StarRocks.g4 (hunk 1 above)
Comment (srlch) [reply fe/fe-core/src/main/java/com/starrocks/sql/parser/StarRocks.g4]: Updated

File: fe/fe-core/src/main/java/com/starrocks/lake/snapshot/ClusterSnapshotMgr.java [hunk 4]
```diff
@@ -0,0 +1,198 @@
...
+
+    public String getLastFinishedAutomatedSnapshotJobName() {
+        ClusterSnapshot snapshot = getAutomatedSnapshot();
+        if (snapshot == null) {
+            return "";
+        }
+        return snapshot.getSnapshotName();
+    }
+
+    public void dropAutomatedSnapshotRequest(boolean isReplay) {
```
Comment (xiangguangyxg) [inline fe/fe-core/src/main/java/com/starrocks/lake/snapshot/ClusterSnapshotMgr.java:137]:
```suggestion
    // Turn off automated snapshot, use stmt for extension in future
    public void setAutomatedSnapshotOff(AdminSetAutomatedSnapshotOffStmt stmt) {
```

File: fe/fe-core/src/main/java/com/starrocks/lake/snapshot/ClusterSnapshotMgr.java [hunk 5]
```diff
@@ -0,0 +1,198 @@
...
+
+    public void load(SRMetaBlockReader reader)
+            throws SRMetaBlockEOFException, IOException, SRMetaBlockException {
+        ClusterSnapshotMgr data = reader.readJson(ClusterSnapshotMgr.class);
+    }
+
+    @Override
+    public void write(DataOutput out) throws IOException {
+        Text.writeString(out, GsonUtils.GSON.toJson(this));
+    }
```
Comment (xiangguangyxg) [inline fe/fe-core/src/main/java/com/starrocks/lake/snapshot/ClusterSnapshotMgr.java:193]: not needed

File: fe/fe-core/src/main/java/com/starrocks/lake/snapshot/ClusterSnapshotMgr.java [hunk 6]
```diff
@@ -0,0 +1,198 @@
...
+    public static final Logger LOG = LogManager.getLogger(ClusterSnapshotMgr.class);
+    public static final String AUTOMATED_NAME_PREFIX = "automated_cluster_snapshot";
+    private static final String UPLOAD_SUB_PATH = "/meta/image/";
+
+    @SerializedName(value = "automatedSnapshotSvName")
+    private String automatedSnapshotSvName = "";
+    @SerializedName(value = "automatedSnapshot")
+    private ClusterSnapshot automatedSnapshot = null;
+    @SerializedName(value = "historyAutomatedSnapshotJobs")
+    private List<ClusterSnapshotJob> historyAutomatedSnapshotJobs = Lists.newArrayList();
```
Comment (xiangguangyxg) [inline fe/fe-core/src/main/java/com/starrocks/lake/snapshot/ClusterSnapshotMgr.java:57]: not thread-safe ?

File: fe/fe-core/src/main/java/com/starrocks/lake/snapshot/ClusterSnapshotMgr.java (hunk 2 above)
Comment (srlch) [reply fe/fe-core/src/main/java/com/starrocks/lake/snapshot/ClusterSnapshotMgr.java]: Updated

File: fe/fe-core/src/main/java/com/starrocks/lake/snapshot/ClusterSnapshotMgr.java (hunk 3 above)
Comment (srlch) [reply fe/fe-core/src/main/java/com/starrocks/lake/snapshot/ClusterSnapshotMgr.java]: Updated

File: fe/fe-core/src/main/java/com/starrocks/lake/snapshot/ClusterSnapshotMgr.java (hunk 4 above)
Comment (srlch) [reply fe/fe-core/src/main/java/com/starrocks/lake/snapshot/ClusterSnapshotMgr.java]: Updated

File: fe/fe-core/src/main/java/com/starrocks/lake/snapshot/ClusterSnapshotMgr.java (hunk 5 above)
Comment (srlch) [reply fe/fe-core/src/main/java/com/starrocks/lake/snapshot/ClusterSnapshotMgr.java]: Updated

File: fe/fe-core/src/main/java/com/starrocks/lake/snapshot/ClusterSnapshotMgr.java (hunk 6 above)
Comment (srlch) [reply fe/fe-core/src/main/java/com/starrocks/lake/snapshot/ClusterSnapshotMgr.java:56]: Updated

File: fe/fe-core/src/main/java/com/starrocks/lake/snapshot/ClusterSnapshotMgr.java [hunk 7]
```diff
@@ -0,0 +1,122 @@
...
+
+    protected void setAutomatedSnapshotOn(AdminSetAutomatedSnapshotStmt stmt, boolean isReplay) {
+        String storageVolumeName = stmt.getStorageVolumeName();
+        if (!isReplay) {
+            ClusterSnapshotLog log = new ClusterSnapshotLog();
+            log.setCreateSnapshotNamePrefix(AUTOMATED_NAME_PREFIX, storageVolumeName);
+            GlobalStateMgr.getCurrentState().getEditLog().logClusterSnapshotLog(log);
+        }
+        automatedSnapshotSvName = storageVolumeName;
+    }
```
Comment (xiangguangyxg) [inline fe/fe-core/src/main/java/com/starrocks/lake/snapshot/ClusterSnapshotMgr.java:58]:
```suggestion
    public void setAutomatedSnapshotOn(AdminSetAutomatedSnapshotStmt stmt) {
        setAutomatedSnapshotOn(stmt.getStorageVolumeName());
    
        ClusterSnapshotLog log = new ClusterSnapshotLog();
        log.setCreateSnapshotNamePrefix(AUTOMATED_NAME_PREFIX, storageVolumeName);
        GlobalStateMgr.getCurrentState().getEditLog().logClusterSnapshotLog(log);
    }

    protected void setAutomatedSnapshotOn(String storageVolumeName) {
        automatedSnapshotSvName = storageVolumeName;
    }
```

File: fe/fe-core/src/main/java/com/starrocks/persist/metablock/SRMetaBlockID.java [hunk 8]
```diff
@@ -97,6 +97,8 @@ public int getId() {
 
     public static final SRMetaBlockID PIPE_MGR = new SRMetaBlockID(32);
 
+    public static final SRMetaBlockID CLOULD_NATIVE_SNAPSHOT_MGR = new SRMetaBlockID(33);
```
Comment (xiangguangyxg) [inline fe/fe-core/src/main/java/com/starrocks/persist/metablock/SRMetaBlockID.java:100]: CLUSTER_SNAPSHOT_MGR

File: fe/fe-core/src/main/java/com/starrocks/sql/ast/AdminSetAutomatedSnapshotStmt.java [hunk 9]
```diff
@@ -0,0 +1,40 @@
...
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package com.starrocks.sql.ast;
+
+import com.starrocks.sql.parser.NodePosition;
+
+public class AdminSetAutomatedSnapshotStmt extends DdlStmt {
```
Comment (xiangguangyxg) [inline fe/fe-core/src/main/java/com/starrocks/sql/ast/AdminSetAutomatedSnapshotStmt.java:19]:
```suggestion
public class AdminSetAutomatedSnapshotOnStmt extends DdlStmt {
```

File: fe/fe-core/src/main/java/com/starrocks/sql/ast/AdminSetOffAutomatedSnapshotStmt.java [hunk 10]
```diff
@@ -0,0 +1,33 @@
...
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package com.starrocks.sql.ast;
+
+import com.starrocks.sql.parser.NodePosition;
+
+public class AdminSetOffAutomatedSnapshotStmt extends DdlStmt {
```
Comment (xiangguangyxg) [inline fe/fe-core/src/main/java/com/starrocks/sql/ast/AdminSetOffAutomatedSnapshotStmt.java:19]:
```suggestion
public class AdminSetAutomatedSnapshotOffStmt extends DdlStmt {
```

File: fe/fe-core/src/main/java/com/starrocks/sql/parser/StarRocks.g4 [hunk 11]
```diff
@@ -734,6 +736,14 @@ syncStatement
     : SYNC
     ;
 
+adminSetAutomatedSnapshotStatement
```
Comment (xiangguangyxg) [inline fe/fe-core/src/main/java/com/starrocks/sql/parser/StarRocks.g4:739]:
```suggestion
adminSetAutomatedSnapshotOnStatement
```

File: fe/fe-core/src/main/java/com/starrocks/sql/parser/StarRocks.g4 [hunk 12]
```diff
@@ -734,6 +736,14 @@ syncStatement
...
+    : ADMIN SET AUTOMATED CLUSTER SNAPSHOT ON (STORAGE VOLUME svName=identifier)?
+    ;
+
+adminSetOffAutomatedSnapshotStatement
```
Comment (xiangguangyxg) [inline fe/fe-core/src/main/java/com/starrocks/sql/parser/StarRocks.g4:743]:
```suggestion
adminSetAutomatedSnapshotOffStatement
```

File: fe/fe-core/src/main/java/com/starrocks/lake/snapshot/ClusterSnapshotMgr.java [hunk 13]
```diff
@@ -0,0 +1,122 @@
...
+    protected void setAutomatedSnapshotOff(AdminSetOffAutomatedSnapshotStmt stmt, boolean isReplay) {
+        if (!isReplay) {
+            ClusterSnapshotLog log = new ClusterSnapshotLog();
+            log.setDropSnapshot(AUTOMATED_NAME_PREFIX);
+            GlobalStateMgr.getCurrentState().getEditLog().logClusterSnapshotLog(log);
+        }
+
+        // drop AUTOMATED snapshot
+        automatedSnapshotSvName = "";
+    }
```
Comment (xiangguangyxg) [inline fe/fe-core/src/main/java/com/starrocks/lake/snapshot/ClusterSnapshotMgr.java:82]: same as above

File: fe/fe-core/src/main/java/com/starrocks/lake/snapshot/ClusterSnapshotMgr.java (hunk 7 above)
Comment (srlch) [reply fe/fe-core/src/main/java/com/starrocks/lake/snapshot/ClusterSnapshotMgr.java]: Updated

File: fe/fe-core/src/main/java/com/starrocks/persist/metablock/SRMetaBlockID.java (hunk 8 above)
Comment (srlch) [reply fe/fe-core/src/main/java/com/starrocks/persist/metablock/SRMetaBlockID.java]: Updated

File: fe/fe-core/src/main/java/com/starrocks/sql/ast/AdminSetAutomatedSnapshotStmt.java (hunk 9 above)
Comment (srlch) [reply fe/fe-core/src/main/java/com/starrocks/sql/ast/AdminSetAutomatedSnapshotStmt.java]: Updated

File: fe/fe-core/src/main/java/com/starrocks/sql/ast/AdminSetOffAutomatedSnapshotStmt.java (hunk 10 above)
Comment (srlch) [reply fe/fe-core/src/main/java/com/starrocks/sql/ast/AdminSetOffAutomatedSnapshotStmt.java]: Updated

File: fe/fe-core/src/main/java/com/starrocks/sql/parser/StarRocks.g4 (hunk 11 above)
Comment (srlch) [reply fe/fe-core/src/main/java/com/starrocks/sql/parser/StarRocks.g4]: Updated

File: fe/fe-core/src/main/java/com/starrocks/sql/parser/StarRocks.g4 (hunk 12 above)
Comment (srlch) [reply fe/fe-core/src/main/java/com/starrocks/sql/parser/StarRocks.g4]: Updated

File: fe/fe-core/src/main/java/com/starrocks/lake/snapshot/ClusterSnapshotMgr.java (hunk 13 above)
Comment (srlch) [reply fe/fe-core/src/main/java/com/starrocks/lake/snapshot/ClusterSnapshotMgr.java]: Updated

**Review by gengjun-git — APPROVED** [review]

**Review by xiangguangyxg — APPROVED** [review]

**Review by kangkaisen — APPROVED** [review]

**Comment by srlch** [comment]
https://github.com/Mergifyio backport branch-3.4

**Pull request merged**